### DISCUSSION TITLE
Security within factory service

### DISCUSSION THREAD

**Discussion by eekes** [symfony/symfony#44520] [Q&A] [answered]
I have a certain service that's created using a factory:

```
App\Service\ExampleInterface:
    factory: ['@App\Factory\ExampleFactory', 'create']
```

Within the ExampleFactory I'm using the Security service, but although I'm logged in, getUser() always returns null. When injecting the Factory directly in a controller (or some other service), getUser() returns the user correctly.

I was wondering why it doesn't work when using it with the "factory" option and if there is a way to get this working?

**Comment by ahundiak** [comment, marked as answer]
This is typically a timing issue.  You are trying to access the user before the security system has a chance to initialize the security system.  If, for example, you inject the factory into an kernel event subscriber and then attempt to use it in the constructor of the subscriber.  Event subscribers are created before the security system does it's thing.  So getUser will fail.  If this is indeed the problem then waiting to use the factory until you actually need whatever it creates will usually do the trick.

If something completely different seems to be happening then perhaps post a few lines of code showing where it fails.

**Reply by eekes** [reply]
That seems to be it. I turned it into a lazy service and now it works fine. 👍
